File: 680. Valid Palindrome II/main.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s) - 1

        while left < right:
            if s[left] != s[right]:
                return self.isEqual(s, left + 1, right) or self.isEqual(
                    s, left, right - 1
                )

            left += 1
            right -= 1

        return True

    def isEqual(self, s: str, left: int, right: int) -> bool:
        while left < right:
            if s[left] != s[right]:
                return False

            left += 1
            right -= 1

        return True


class Solution:
    def validPalindrome(self, s: str) -> bool:
        return self.isEqual(s, 0, len(s) - 1, False)

    def isEqual(self, s: str, left: int, right: int, skipped: bool) -> bool:
        while left < right:
            if s[left] != s[right]:
                if skipped:
                    return False
                return self.isEqual(s, left + 1, right, True) or self.isEqual(
                    s, left, right - 1, True
                )

            left += 1
            right -= 1

        return True


class Solution:
    def validPalindrome(self, s: str) -> bool:
        return self.isEqual(s, 0, len(s) - 1, False)

    def isEqual(self, s: str, left: int, right: int, skipped: bool) -> bool:
        if left >= right:
            return True

        if s[left] != s[right]:
            if skipped:
                return False
            return self.isEqual(s, left + 1, right, True) or self.isEqual(
                s, left, right - 1, True
            )

        return self.isEqual(s, left + 1, right - 1, skipped)


class Solution:
    def validPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s) - 1

        while left < right:
            if s[left] != s[right]:
                skipR = s[left + 1 : right + 1]
                skipL = s[left:right]

                return skipR == skipR[::-1] or skipL == skipL[::-1]

            left += 1
            right -= 1

        return True

File: 680. Valid Palindrome II/main.rs
struct Solution;

impl Solution {
    pub fn valid_palindrome(s: String) -> bool {
        let chars = s.chars().collect::<Vec<_>>();
        let mut left = 0;
        let mut right = chars.len() - 1;

        while left < right {
            if chars[left] != chars[right] {
                let skip_r = chars[left + 1..right + 1].into_iter().collect::<Vec<_>>();
                let skip_l = chars[left..right].into_iter().collect::<Vec<_>>();
                let rev_r = skip_r.clone().into_iter().rev().collect::<Vec<_>>();
                let rev_l = skip_l.clone().into_iter().rev().collect::<Vec<_>>();
                return skip_r.eq(&rev_r) || skip_l.eq(&rev_l);
            }

            left += 1;
            right -= 1;
        }

        true
    }
}

impl Solution {
    pub fn valid_palindrome(s: String) -> bool {
        let chars = s.chars().collect::<Vec<_>>();
        let mut left = 0;
        let mut right = chars.len() - 1;

        while left < right {
            if chars[left] != chars[right] {
                return Solution::is_equal(&chars, left + 1, right)
                    || Solution::is_equal(&chars, left, right - 1);
            }

            left += 1;
            right -= 1;
        }

        true
    }

    pub fn is_equal(chars: &Vec<char>, mut left: usize, mut right: usize) -> bool {
        while left < right {
            if chars[left] != chars[right] {
                return false;
            }

            left += 1;
            right -= 1;
        }

        true
    }
}

impl Solution {
    pub fn valid_palindrome(s: String) -> bool {
        let chars = s.chars().collect::<Vec<_>>();
        return Solution::is_equal(&chars, 0, chars.len() - 1, false);
    }

    pub fn is_equal(chars: &Vec<char>, mut left: usize, mut right: usize, skipped: bool) -> bool {
        while left < right {
            if chars[left] != chars[right] {
                if skipped {
                    return false;
                }
                return Solution::is_equal(chars, left + 1, right, true)
                    || Solution::is_equal(chars, left, right - 1, true);
            }

            left += 1;
            right -= 1;
        }

        true
    }
}

impl Solution {
    pub fn valid_palindrome(s: String) -> bool {
        let chars = s.chars().collect::<Vec<_>>();
        return Solution::is_equal(&chars, 0, chars.len() - 1, false);
    }

    pub fn is_equal(chars: &Vec<char>, mut left: usize, mut right: usize, skipped: bool) -> bool {
        if left >= right {
            return true;
        }

        if chars[left] != chars[right] {
            if skipped {
                return false;
            }
            return Solution::is_equal(chars, left + 1, right, true)
                || Solution::is_equal(chars, left, right - 1, true);
        }

        return Solution::is_equal(chars, left + 1, right - 1, skipped);
    }
}
